streaming/base/converters/dataframe_to_mds.py

```diff
@@ -0,0 +1,301 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""A utility to convert spark dataframe to MDS."""
+
+import json
+import logging
+import os
+import shutil
+from argparse import ArgumentParser, Namespace
+from collections.abc import Iterable
+from typing import Any, Callable, Dict, Iterable, Optional, Tuple, Union
+
+import pandas as pd
+from pyspark import TaskContext
+from pyspark.sql import SparkSession
+from pyspark.sql.dataframe import DataFrame
+from pyspark.sql.types import (ArrayType, BinaryType, BooleanType, ByteType, DateType,
+                               DayTimeIntervalType, DecimalType, DoubleType, FloatType,
+                               IntegerType, LongType, MapType, ShortType, StringType, StructField,
+                               StructType, TimestampNTZType, TimestampType)
+
+from streaming import MDSWriter
+from streaming.base.format.index import get_index_basename
+from streaming.base.format.mds.encodings import _encodings
+from streaming.base.storage.upload import CloudUploader
+
+logger = logging.getLogger(__name__)
+
+
+def is_iterable(obj: Any) -> bool:
+    """Check if obj is iterable."""
+    return issubclass(type(obj), Iterable)
+
+
+def infer_dataframe_schema(dataframe: DataFrame,
+                           user_defined_cols: Optional[Dict[str, Any]] = None) -> Optional[Dict]:
+    """Retrieve schema to construct a dictionary or do sanity check for MDSWriter.
+
+    Args:
+        dataframe (spark dataframe): dataframe to inspect schema
+        user_defined_cols (dict): user specified schema for MDSWriter
+    Return:
+        If user_defined_cols is None, return schema_dict (dict): column name and dtypes that supported by MDSWriter
+        Else, return None
+    Exceptions:
+        Any of the datatype found to be unsupported by MDSWriter, then raise ValueError
+    """
+    dtype_mapping = {
+        ByteType: 'bytes',
+        ShortType: 'uint64',
+        IntegerType: 'int',
+        LongType: 'int64',
+        FloatType: 'float32',
+        DoubleType: 'float64',
+        DecimalType: 'str_decimal',
+        StringType: 'str',
+        BinaryType: None,
```

Comment:
```suggestion
        BinaryType: 'bytes',
```

Reply:
thought it should be MDS' `uint8` type

Reply:
@knighton can you confirm that? We thought BinaryType is a byte array, so it should map to bytes.

Reply:
@knighton You have it backwards I think. ByteType should be a uint8.

Reply:
Yes, uint8 makes sense. thanks~